Type cleanup

scober opened the pull request:
Unfortunately, `PropertyGraph` just doesn't have enough context to perform all of its own loading logic. So this PR adds some functionality to `PropertyGraphRetractor` to help out the classes that help load `PropertyGraph`s. It propagates information about whether or not types need to be inferred from properties and it exposes direct access to the underlying type id arrays because the pointer is ugly and this isn't going away any time soon.

This PR also changes the logic for inferring types to only treat `uint8` properties as types (and not `boolean` types). The boolean types are not widely used and they pose a problem because Cypher allows boolean properties but we are treating every boolean property as a type.

ddn0 commented on libquery/src/MatchOp.cpp:
```diff
@@ -131,15 +131,59 @@ galois::MatchOp::MakeNoResult() const {
+}
 std::vector<galois::Operator*>
 galois::MatchOp::Compute() {
+  // XXX(sam) current task: determine which variables are already bound to this.
```
`XXX` => `TODO`

https://github.com/KatanaGraph/katana/blob/master/CONTRIBUTING.md#code-style-and-conventions

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -730,7 +735,50 @@ class SubgraphQueryMiner {
     return QueryEmbeddingList(std::move(concurrent_list));
   }
+  /**
+   * 
```
???

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -730,7 +735,50 @@ class SubgraphQueryMiner {
+          local_list.push_back(emb);
+        });
+    return QueryEmbeddingList(std::move(concurrent_list));
+  }
```
newlines between function definitions please

ddn0 commented on libquery/src/MatchOp.cpp:
```diff
@@ -131,15 +131,59 @@ galois::MatchOp::MakeNoResult() const {
 
+const std::shared_ptr<galois::Table>
+galois::MatchOp::GetStartingTable() const {
+  if (this->GetNumInputs() == 0) {
```
probably just `GetNumInputs` without `this->`

ddn0 commented on libquery/src/MatchOp.cpp:
```diff
@@ -131,15 +131,59 @@ galois::MatchOp::MakeNoResult() const {
+    //make a 1-entry empty table
+    const std::vector<TableNodeName> no_nodes;
+    const std::vector<TableEdgeName> no_edges;
+    return std::shared_ptr<Table>(Table::MakeEmpty(no_nodes, no_edges));
```
maybe provide an overload for `MakeEmpty` for a completely empty table with no schema whatsoever?

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -811,13 +874,15 @@ galois::SubgraphQuery(
     const QueryEdgeProperties& query_edge_props,
     const graphs::AttributedGraph& attributed_graph,
-    const MatchOrder& match_order) {
+    const MatchOrder& match_order, const std::shared_ptr<Table> input_table,
```
`const std::shared_ptr<Table>&`

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -750,12 +798,15 @@ class SubgraphQueryMiner {
       const graphs::AttributedGraph& attributed_graph, QueryGraph& query_graph,
       const QueryNodeProperties& query_node_props,
-      const MatchOrder& match_order)
+      const MatchOrder& match_order, const std::shared_ptr<Table> input_table,
```
`const std::shared_ptr<Table>&`

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -750,12 +798,15 @@ class SubgraphQueryMiner {
       const QueryNodeProperties& query_node_props,
-      const MatchOrder& match_order)
+      const MatchOrder& match_order, const std::shared_ptr<Table> input_table,
+      const std::map<unsigned, unsigned> prematched_indices)
```
`const std::map<unsigned, unsigned>&`

ddn0 commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -772,20 +823,32 @@ class SubgraphQueryMiner {
+      list = AddNode(GetQueryVertex(0));
+      match_level = 1;
+    } else {
+      GALOIS_LOG_DEBUG(
```
probably remove this debugging print code?

ddn0 commented on libquery/src/MatchOp.cpp:
```diff
@@ -391,37 +435,42 @@ class HighestDegreeFirstOrder {
+  std::set<VertexId> inserted(start.begin(), start.end());
+  VertexId start_index = start.size();
+  if (start_index == 0) {
+    //have to actually choose a starting vertex
```
? `// choose a starting vertex`

at least `//<space>`  throughout.

ddn0 commented on libquery/include/querying/SubgraphQuery.h:
```diff
@@ -12,5 +12,6 @@ QueryEmbeddingList SubgraphQuery(
     const QueryEdgeProperties& query_edge_props,
     const graphs::AttributedGraph& attributed_graph,
-    const MatchOrder& matching_order);
+    const MatchOrder& matching_order, const std::shared_ptr<Table> input_table,
```
`const T&`

ddn0 reviewed (commented):
small style nits

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -731,6 +736,46 @@ class SubgraphQueryMiner {
+    //predetermine
+    galois::do_all(
+        //iterate over the table
+        galois::iterate(size_t{0}, input_matches_->NumRows()),
```
Is this input table distributed among hosts?

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -772,20 +820,25 @@ class SubgraphQueryMiner {
+    // Number of (presumably) matched nodes of query graph; start at 1 since
+    // above do_all matches (or doesn't match) a single node; in no match case,
+    // list will be empty anyways
+    unsigned match_level = 1;
```
Better to not initialize here and let the if-else below handle it. Then the comments are not needed.

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -772,20 +820,25 @@ class SubgraphQueryMiner {
+    }
     if (query_graph_.size() <= 1) {
       if constexpr (kPrintEmbedding) {
         PrintEmbedding(list, 0);
```
`PrintEmbedding(list, match_level - 1)`

roshandathathri commented on lonestar/querying/distributed/graph-query/cypher-queries/person-company-university-city-country-with-with.q:
```diff
@@ -0,0 +1,4 @@
+match (p:Person)-[:WORK_AT]->(:Company)-[:IS_LOCATED_IN]->(c:Country)
+with p,c
+match (p)-[:STUDY_AT]->(:University)-[:IS_LOCATED_IN]->(:City)-[:IS_PART_OF]->(c)
+return count(p);
```
Please add newline at the end of the file. You can configure your editor to do that as well.

Also, just to do be consistent with other query file, please use caps for the keywords MATCH, WITH, RETURN.

swreinehr replied:
I just based it off of the person-company-university-city-country.q, which had them lowercase. There are several other queries that are lowercase.

Also, if we want to standardize the queries it would be convenient to do them all at once in another PR, where we also establish the format. and add newlines to make them more readable. Should I open an issue for this?

roshandathathri replied:
Yes, sounds good.

Please ensure newlines in this commit though.

roshandathathri reviewed (commented):
Please add a couple more tests - preferably those that follow from some LDBC IC queries.

roshandathathri commented on libquery/src/MatchOp.cpp:
```diff
@@ -133,12 +133,67 @@ galois::MatchOp::MakeNoResult() const {
+    auto index = var_name_to_node_index_.find(name);
+    if (index != var_name_to_node_index_.end()) {
+      result[index->second] = col;
+      GALOIS_LOG_DEBUG(
```
Drop this when you merge.

roshandathathri commented on libquery/src/Table.cpp:
```diff
@@ -142,6 +142,117 @@ galois::Table::Make(
+        GALOIS_LOG_ASSERT(st.ok());
+        ++v_col;
+      }
+      //! TODO (danielmawhirter) do not use AppendUnsafe, size is not known
```
Is this relevant? Don't see `AppendUnsafe`

swreinehr replied:
that's a comment that was copied from the method above that I modified since size wouldn't be known. I was thinking about combining those two methods, but was not sure about it. I've changed the comment

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -71,6 +72,12 @@ class SubgraphQueryMiner {
+  /// map of nodes that are already matched, and should be skipped, and what position in the original table they had
+  const std::map<unsigned, unsigned> prematched_indices_;
+  //! mapping to input table, part of the output
+  std::map<QueryEmbedding, std::vector<size_t>>& row_mapping_;
```
Pointer instead of reference as it is being modified.

I would prefer if the SubgraphQuery interface takes an embedding list as input. The table to embedding list (and the row mapping) can be done before calling SubgraphQuery, right?

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -731,6 +738,57 @@ class SubgraphQueryMiner {
+              return;
+            }
+            emb.push_back(node_id);
+            if (!attributed_graph_.MatchNodeDegree(
```
Confused by this. Shouldn't this be before pushing node_id onto emb?

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -731,6 +738,57 @@ class SubgraphQueryMiner {
+          // make sure the mapping exists
+          // if it does not, create the mapping
+          // and insert the embedding into the list.
+          // need to mutex this
```
It would be a global mutex. Better to do this on a single thread instead.

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -772,20 +835,25 @@ class SubgraphQueryMiner {
-    QueryEmbeddingList list = AddNode(GetQueryVertex(0));
+    // Number of (presumably) matched nodes of query graph; start at 1 since
+    // above do_all matches (or doesn't match) a single node; in no match case,
+    // list will be empty anyways
```
Drop/rectify the comment

Commit:
Propogate IsEntityTypeIDsOutsideOfProperties()

This function is used whenever a graph is loaded. Unfortunately,
PropertyGraph doesn't have the context to perform all graph-loading
logic (think testing, distributed execution, etc.). So this function is
propogated through to PropertyGraphRetractor so that classes that futz
with PropertyGraphs have this information.

Commit:
Only treat uint8 properties as types

We currently treat uint8 and boolean properties as types. But boolean
properties aren't widely used as types and boolean properties are
allowed in Cypher (while uint8 properties are not), so we eliminate
the boolean option before it causes problems.

amberhassaan commented on libgalois/include/katana/PropertyGraphRetractor.h:
```diff
@@ -69,10 +69,30 @@ class KATANA_EXPORT PropertyGraphRetractor {
 
+  /// Return true if type information has been loaded separate from properties.
+  /// Return false otherwise.
+  bool IsEntityTypeIDsOutsideProperties() {
```
We could simplify this name to `hasEntityTypeIDs` or `hasValidEntityTypeIDs`

scober replied:
I would love to. But I think we should make clear that this function is checking for whether or not the types are expected to be inferrable from the `uint8` properties, which is not exactly the same as whether there are valid types present in the graph.

scober replied:
Right now I am going with `NeedsEntityTypeIDInference()`

amberhassaan replied:
That's OK too.

amberhassaan commented on libgalois/include/katana/PropertyGraphRetractor.h:
```diff
@@ -69,10 +69,30 @@ class KATANA_EXPORT PropertyGraphRetractor {
+
+  /// Modifiable access to the underlying type array.
+  /// This is exposed because type id mappings change sometimes.
+  PropertyGraph::EntityTypeIDArray& NodeTypeArray() {
```
I would still favor returning a pointer to the NUMAArray buffer as opposed to reference to NUMAArray. That way, at least ownership of memory is known.

scober replied:
Why does the reference not make ownership of the memory clear?

witchel replied:
When I see a C pointer I know I don't own the memory.  Mutable references are frowned on (by clang-tidy and I think our coding standard).  It is easy to copy them without realizing it.  At least for me.

scober replied:
Okay. Could I do a pointer to the array and not the data buffer? (pointer arithmetic is also frowned upon by clang-tidy)

amberhassaan replied:
NO, that's worse because that makes it even easier to do bad things to the NUMAArray. A pointer to buffer hides the fact that it was a NUMAArray internally. and like Emmett said, it conveys the message that you don't own the memory.

scober replied:
Well then I would like to vote in favor of a buffer class again.

amberhassaan replied:
I'm gonna implement it any day now. and Arrow fans will tell me that "Arrow has a Buffer already".

amberhassaan commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -137,12 +137,6 @@ class KATANA_EXPORT EntityTypeManager {
     // assign the type ID for each row
     for (int64_t row = 0; row < num_rows; ++row) {
       TypeProperties::FieldEntity field_indices;
-      for (auto& bool_property : type_properties.bool_properties) {
```
So it's only uint8 properties now?

scober replied:
Yes. @roshandathathri says we don't need to support boolean properties any more.

amberhassaan reviewed (approved):
Minor comments but overall OK.

witchel commented on libgalois/include/katana/PropertyGraphRetractor.h:
```diff
@@ -69,10 +69,30 @@ class KATANA_EXPORT PropertyGraphRetractor {
+
+  /// This is exposed because type id mappings change sometimes.
   void ReplaceNodeTypeManager(EntityTypeManager&& manager) {
     pg_->node_entity_type_manager_ = std::move(manager);
```
I wonder if we should have a valid bit in the EntityTypeManager, so if you use one after it being moved, it barfs all over you.

scober replied:
We already have an `Init()` so there is a logical place to set the bit. I would be on board.

witchel replied:
I'm on it.

amberhassaan replied:
I personally am not a fan of valid bit but I realize they are useful sometimes. valid bits are good until they aren't. IMO, it's better to deduce validity through the state of the object, e.g. empty maps mean EntityTypeManager is invalid because a valid one would have at least `kUnknownType` mapped.

witchel replied:
Well, I implemented it in my branch, so let's discuss before it becomes a PR.  What is ugly about the valid bit is that it needs to be checked everywhere.  But it provides fast fail for when an object has been moved.  This happens in graph update, where I move entity type managers and I've certainly coded the bug where I use it after moving it.  But maybe this problem isn't important enough to write and maintain a valid bit.

Amber, no "empty behavior" is sufficient.  I want to fail an assert if I try to use an invalid manager.

amberhassaan replied:
I'd like to see your code that has these moves and related bugs (or possibility of one). I also still don't understand why we can't implement `isValid()` by checking for empty maps? A "default/no-arg" constructed manager has non-empty maps, and a moved one should have empty maps.

witchel replied:
Moved discussion to slack.
https://thek-teamworkspace.slack.com/archives/C028THYNA07/p1631379992027200

witchel commented on libsupport/src/EntityTypeManager.cpp:
```diff
@@ -24,15 +24,9 @@ katana::EntityTypeManager::DoAssignEntityTypeIDsFromProperties(
-    } else if (current_field->type()->Equals(arrow::uint8())) {
+    if (current_field->type()->Equals(arrow::uint8())) {
       type_field_indices.push_back(i);
       std::shared_ptr<arrow::Array> property = properties->column(i)->chunk(0);
```
I'm probably overly paranoid about this issue, but I might assert (debug?) that there is only one chunk.

scober replied:
Will do

witchel reviewed (commented):
Looks good.
